hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/DefaultCertificateClient.java

```diff
@@ -1095,4 +1089,274 @@ public synchronized void close() throws IOException {
+        File currentCertDir = new File(
+            securityConfig.getCertificateLocation(component).toString());
+        File backupKeyDir = new File(
+            securityConfig.getKeyLocation(component).toString() +
+                HDDS_BACKUP_KEY_CERT_DIR_NAME_SUFFIX);
+        File backupCertDir = new File(
+            securityConfig.getCertificateLocation(component).toString() +
+                HDDS_BACKUP_KEY_CERT_DIR_NAME_SUFFIX);
+
+        if (!currentKeyDir.renameTo(backupKeyDir)) {
+          // Cannot rename current key dir to the backup dir
+          throw new CertificateException("Failed to rename " +
+              currentKeyDir.getAbsolutePath() +
+              " to " + backupKeyDir.getAbsolutePath() + " during " +
+              "certificate renew.", RENEW_ERROR);
+        }
+        if (!currentCertDir.renameTo(backupCertDir)) {
+          // Cannot rename current cert dir to the backup dir
+          rollbackDir(currentKeyDir, currentCertDir, newKeyDir, newCertDir,
+              backupKeyDir, backupCertDir, "step-1");
+          throw new CertificateException("Failed to rename " +
+              currentCertDir.getAbsolutePath() +
+              " to " + backupCertDir.getAbsolutePath() + " during " +
+              "certificate renew.", RENEW_ERROR);
+        }
+
+        if (!newKeyDir.renameTo(currentKeyDir)) {
+          // Cannot rename new dir as the current dir
+          String msg = "Failed to rename " + newKeyDir.getAbsolutePath() +
+              " to " + currentKeyDir.getAbsolutePath() +
+              " during certificate renew.";
+          // rollback
+          rollbackDir(currentKeyDir, currentCertDir, newKeyDir, newCertDir,
+              backupKeyDir, backupCertDir, "step-2");
+          throw new CertificateException(msg, RENEW_ERROR);
+        }
+
+        if (!newCertDir.renameTo(currentCertDir)) {
+          // Cannot rename new dir as the current dir
+          String msg = "Failed to rename " + newCertDir.getAbsolutePath() +
+              " to " + currentCertDir.getAbsolutePath() +
+              " during certificate renew.";
+          // rollback
+          rollbackDir(currentKeyDir, currentCertDir, newKeyDir, newCertDir,
+              backupKeyDir, backupCertDir, "step-2");
+          throw new CertificateException(msg, RENEW_ERROR);
+        }
+
+        // Delete backup dir on next DN startup
+        getLogger().info("Successful renew key and certificate." +
+            " New certificate {}.", dnCertSerialId);
+        return dnCertSerialId;
+      } finally {
+        isRenewing.set(false);
+      }
+    }
+
+    throw new CertificateException("A renewAndStoreKeyAndCert process" +
+        " is running already.", RENEW_ERROR);
+  }
+
+  private void rollbackDir(File currentKeyDir, File currentCertDir,
+      File newKeyDir, File newCertDir, File backupKeyDir, File backupCertDir,
+      String step) throws CertificateException {
```

Comment:
The step parameter is a bit interesting for me here, and I would propose to drop it the following way and for the following reasons:
We fail the whole process and throw an exception, when:
- if cleanup of the new directory fails
- if creating the new new directory fails
- if cleanup the backup directory fails
- if creating the keypair into the new directory fails
- if creating and signing the CSR fails
- if moving the current key dir to backup fails
Other than this there is a rollback called.

I would argue that as the whole process starts with removing the new dir if exists, we do not need to move the current back to new during the rollback, we just need to ensure that the backed up data is moved back to the actual location where current data should be.

If we change to this approach, then there is one more thing we can do, after the new data moved to the current location and successfully applied, we should also remove the backup location.
So that the next run, we do not need to clean the backup location, as if it is there, that would mean that a rollback has failed, or the last step (removing the backup) failed when applying a new certificate, in this case we can fail, and require a human check to get out of the situation.

I am not sure it is clear, but after the suggested changed the simplified flow would be:
- start the process, clear new folder (just in case)
- assert that backup location does not exists, if it does exists than fail and ask human help
- create the new new folder, key and CSR, and then sign the cert and save it to the new dir
- move the current folder to backup location (if fails rollback and move back what has been moved)
- move the new folder to current location (if fails, rollback and move back the backup to current)
- clear the backup location (the current remains the only directory)
If during this process the rollback fails, then just log an error, next run will fail on asserting that backup locations do not exists.

Reply:
@fapifta , one thing can improve is we may not need to move current back to new if move new folder to current location failed. 

Since we are doing the key and cert renew far before they expire, for some recoverable failures, such as certificate sign request sent to SCM who cannot response in time due to any unexpected reason, we should tolerate this case and have renew retry later. 

So I defined two CertificateException error code, one for renew error, one for rollback error. Renew error means something is wrong with the key and cert renew but it doesn't have bad impact to current key and certs, so the service can continue run, for example, the above case. 

Rollback error means there is something happened which requires admin manual intervention, such as fail to move backup to current location. In this case, the monitor task will terminate the service process.

Reply:
Got it, this seems to be much much better. Thank you for taking the extra mile to find a better way of expressing the difference.